.github/workflows/ci-tests.yml

```diff
@@ -1,7 +1,211 @@
+        echo "summary<<${delimiter}" >> $GITHUB_OUTPUT
+        echo "## Terraform Plan Output" >> $GITHUB_OUTPUT
+        echo "<details><summary>Click to expand</summary>" >> $GITHUB_OUTPUT
+        echo "" >> $GITHUB_OUTPUT
+        echo '```terraform' >> $GITHUB_OUTPUT
+        echo "$TERRAFORM_PLAN" >> $GITHUB_OUTPUT
+        echo '```' >> $GITHUB_OUTPUT
+        echo "</details>" >> $GITHUB_OUTPUT
+        echo "${delimiter}" >> $GITHUB_OUTPUT
+      
+    # Publish Terraform Plan as task summary
+    - name: Publish Terraform Plan to Task Summary
+      if: github.event.pull_request.state != 'closed'
+      env:
+        SUMMARY: ${{ steps.tf-plan-string.outputs.summary }}
+      run: |
+        echo "$SUMMARY" >> $GITHUB_STEP_SUMMARY
+
+    # If this is a PR post the changes
+    - name: Push Terraform Output to PR
+      if: github.ref != 'refs/heads/main' && github.event.pull_request.state != 'closed'
+      uses: actions/github-script@v7
+      env:
+        SUMMARY: "${{ steps.tf-plan-string.outputs.summary }}"
+      with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          script: |
+            const body = `${process.env.SUMMARY}`;
+            github.rest.issues.createComment({
+                issue_number: context.issue.number,
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                body: body
+            })
+            
+    - name: Error in Terraform 
+      if: steps.tf-plan.outputs.exitcode == 1  && github.event.pull_request.state != 'closed'
+      run: exit 1
+
+    # Terraform Apply if no errors in plan & changes detected 
+    - name: Terraform Apply 
+      id: tf-apply    
+      if: steps.tf-plan.outputs.exitcode == 2 && github.event.pull_request.state != 'closed'
+      run: |
+        terraform apply -auto-approve tfplan       
+    
+    # Set environment variables for data validation from terraform output 
+    - name: Set Data Validation Variables 
+      id: prepare-data-validation
+      if:  steps.tf-apply.outcome != 'failure' && github.event.pull_request.state != 'closed'
+      run: |
+        terraform_output=$(terraform output -json)  
+        observe_token_id=$(echo $terraform_output | jq -r '.observe_token_id.value')
+        azure_dataset_id=$(echo $terraform_output | jq -r '.azure_dataset_id.value')
+        echo "OBSERVE_TOKEN_ID=$observe_token_id" >> $GITHUB_ENV
+        echo "AZURE_DATASET_ID=$azure_dataset_id" >> $GITHUB_ENV
+        
+    - name: Data Validation Test 
+      uses: nick-fields/retry@v3
+      if:  steps.tf-apply.outcome != 'failure' && github.event.pull_request.state != 'closed'
+      with:
+        timeout_minutes: 30
+        max_attempts: 40 #Max 40 attempts (resource data can take longer depending on when the next cron job runs)
+        retry_wait_seconds: 60  #1 attempt/minute
+        command: |
+            pip install requests && python ${{github.workspace}}/.github/scripts/query_observe.py
```

Comment:
You should manage python dependencies with a requirements.txt file

Reply:
Usually I would but in this case it's just one library and rest is all python native libs.